modules/frontend/combiner/span_matcher.go

```diff
@@ -0,0 +1,247 @@
+package combiner
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/grafana/tempo/pkg/spanfilter/config"
+	"github.com/grafana/tempo/pkg/spanfilter/policymatch"
+
+	"github.com/grafana/tempo/pkg/tempopb"
+	commonv1 "github.com/grafana/tempo/pkg/tempopb/common/v1"
+	v1 "github.com/grafana/tempo/pkg/tempopb/resource/v1"
+	tracev1 "github.com/grafana/tempo/pkg/tempopb/trace/v1"
+	"github.com/grafana/tempo/pkg/traceql"
+)
+
+const SpanMatcherHeader = "X-Span-Matcher"
+
+type SpanMatcher struct {
+	Policies []*FilterPolicy
+}
+
+type FilterPolicy struct {
+	Matchers []*PolicyMatcher
+}
+
+type PolicyMatcher struct {
+	shouldMatch     bool
+	spanFilter      *policymatch.AttributePolicyMatch
+	resourceFilter  *policymatch.AttributePolicyMatch
+	intrinsicFilter *policymatch.IntrinsicPolicyMatch
+}
+
+func (fp *FilterPolicy) MatchIntrinsic(span *tracev1.Span) bool {
+	// these are AND operators between matchers so exit early if any matcher does not match
+	for _, matcher := range fp.Matchers {
+		if matcher.intrinsicFilter == nil {
+			continue
+		}
+		if matcher.shouldMatch != matcher.intrinsicFilter.Matches(span) {
+			return false
+		}
+	}
+	return true
+}
+
+func (fp *FilterPolicy) MatchSpan(span *tracev1.Span) bool {
+	// these are AND operators between matchers so exit early if any matcher does not match
+	for _, matcher := range fp.Matchers {
+		if matcher.spanFilter == nil {
+			continue
+		}
+		if matcher.shouldMatch != matcher.spanFilter.Matches(span.Attributes) {
+			return false
+		}
+	}
+	return true
+}
+
+func (fp *FilterPolicy) MatchResource(rsAttrs []*commonv1.KeyValue) bool {
+	// these are AND operators between matchers so exit early if any matcher does not match
+	for _, matcher := range fp.Matchers {
+		if matcher.resourceFilter == nil {
+			continue
+		}
+		if matcher.shouldMatch != matcher.resourceFilter.Matches(rsAttrs) {
+			return false
+		}
+	}
+	return true
+}
+
+func NewSpanMatcher(matcherValue string) (*SpanMatcher, error) {
+	if matcherValue == "" {
+		return nil, errors.New("span matcher header value is empty")
+	}
+	matcherValue = strings.TrimSpace(matcherValue)
```

Comment:
Not in love with all this manual string mangling but if that is the best way then :shrug: .